ESLint: Add rule to prevent usage of the word 'sidebar' in translatable strings

im3dabasia opened the pull request:
Related comment: https://github.com/WordPress/gutenberg/issues/61499#issuecomment-2607544219

## What?
This PR introduces a new ESLint rule to prevent the use of the term "sidebar" in user-facing strings throughout the codebase. The rule aims to ensure that more generic and accessible terms like "panel" are used in place of "sidebar."


## Why?
The term "sidebar" is problematic for blind and low-vision users, as it relies on spatial concepts that are irrelevant to them. Additionally, the term can be misleading on smaller screens where "sidebars" are often full-width overlays. This change is necessary to align with the accessibility guidelines. It ensures consistency in user-facing language across WordPress projects.


## How?
The PR implements an ESLint rule that flags occurrences of the word "sidebar" in translatable strings, helping developers avoid its usage in user-facing content. The rule is added to the .eslintrc.js configuration to prevent new instances of this term from being introduced.


## Testing Instructions

1. Change any word inside any translation ready functions to "sidebar" in any file.
2. Check if your IDE raises an ESLint error with the message: "Avoid using the word 'sidebar' in translatable strings."


## Screencast 


https://github.com/user-attachments/assets/42dcf1b3-315d-43cb-b798-954a8030c4c4

im3dabasia committed:
add: Sidebar lint warning in eslintrc

im3dabasia committed:
fix: Add eslint rule to ensure word is flagged at all positions

im3dabasia committed:
fix: Update regex for better control

im3dabasia commented:
Hi @swissspidy,

When you have a moment, could you please review my PR.

cc: @afercia

tyxla commented on .eslintrc.js:
```diff
@@ -119,6 +119,11 @@ const restrictedSyntax = [
 			'CallExpression[callee.name=/^(__|_x|_n|_nx)$/] > Literal[value=/^toggle\\b/i]',
 		message: "Avoid using the verb 'Toggle' in translatable strings",
 	},
+	{
+		selector:
+			'CallExpression[callee.name=/^(__|_x|_n|_nx)$/] > Literal[value=/sidebar\\b/i]',
+		message: "Avoid using the word 'sidebar' in translatable strings",
+	},
```
As these undesired terms grow, we might want to have an ESLint rule that supports multiple of them.

afercia replied:
I would suggest 'panel' as alternative.

im3dabasia replied:
Thank you, @afercia. I have updated the message to suggest 'panel' as an alternative word.

tyxla commented on .eslintrc.js:
```diff
@@ -119,6 +119,11 @@ const restrictedSyntax = [
+	{
+		selector:
+			'CallExpression[callee.name=/^(__|_x|_n|_nx)$/] > Literal[value=/sidebar\\b/i]',
+		message: "Avoid using the word 'sidebar' in translatable strings",
```
Should we recommend alternatives right there in the ESLint rule message?

im3dabasia replied:
Yeah, that's a good idea. I have updated the message to include a suggestion.

tyxla reviewed (commented):
Looks good overall 👍

im3dabasia committed:
fix: Update rule message to include alternate word suggestion

im3dabasia commented:
The lint checks are failing, but the ones related to this PR possibly include... I'll put together a comprehensive set of suggestions so we can move forward.

https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/edit-site/src/components/add-new-template/add-custom-generic-template-modal-content.js#L45-L56


https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/edit-site/src/components/welcome-guide/page.js#L53-L64

https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/edit-widgets/src/components/sidebar/widget-areas.js#L34-L38

https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/editor/src/components/page-attributes/parent.js#L279-L294

https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/editor/src/components/post-excerpt/index.js#L78-L90

https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/editor/src/components/post-template/create-new-template-modal.js#L120-L131

https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/editor/src/components/post-url/index.js#L90-L106

https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/editor/src/components/preferences-modal/index.js#L87-L94


https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/editor/src/components/sidebar/header.js#L19-L27

https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/editor/src/components/sidebar/index.js#L104-L107

https://github.com/WordPress/gutenberg/blob/be0f32a4498a1cc6c4fab1a82fe06f5c6ecd6eae/packages/fields/src/fields/parent/parent-edit.tsx#L323-L337

im3dabasia committed:
fix: Updated strings to use panel

im3dabasia committed:
fix: Update sidebar lint regex

im3dabasia committed:
fix: Update string to use side bar in sidebar

im3dabasia commented:
Hi @afercia , @swissspidy , @tyxla ,

I’d appreciate your input once again in suggesting alternative strings for the following:

For now, I have replaced "sidebar" with "panel," but if you have any alternative suggestions for any of these, I’d be happy to update them.

1.  'Describe the template, e.g. "Post with **sidebar**". A custom template can be manually applied to any post or page.' 
2.  'It’s now possible to edit page content in the site editor. To customise other parts of the page like the header and footer switch to editing the template using the settings **sidebar**.' 
3. 'Widget Areas are global parts in your site’s layout that can accept blocks. These vary by theme, but are typically parts like your **Sidebar** or Footer.' 
4. 'Describe the template, e.g. "Post with **sidebar**". A custom template can be manually applied to any post or page.' 
5. 'Opens the List View **sidebar** by default.'
6. _x( 'Document', 'noun, **sidebar**' )
7. _x( 'Settings', '**sidebar** button label' ) 

Additionally, we have URLs inside translation functions( 4 instances ), such as:

__('https://wordpress.org/documentation/article/page-post-settings-sidebar/#page-attributes')

We could add an ESLint ignore here, but I don’t think that’s the best solution. A better approach would be to update the regex.

The previous regex worked well, but it caused false positives for URLs because \b only checks for alphanumeric characters, and a hyphen (-) is not considered alphanumeric. This led to unintended matches.

I have updated the regex to ensure that "sidebar" is only flagged when it is a standalone word and not part of another word (e.g., "no-sidebar").

Updated regex:


```json
{
	selector:
		'CallExpression[callee.name=/^(__|_x|_n|_nx)$/] > Literal[value=/(?<![-\\w])sidebar(?![-\\w])/i]',
	message:
		"Avoid using the word 'sidebar' in translatable strings. Consider using 'panel' instead.",
}
```
I have updated the PR with this new regex. Let me know your thoughts! @swissspidy

swissspidy commented:
New regex looks good at first glance, seems to nicely prevent false positives.

I don't think we should change the instances where `sidebar` refers to a theme template part ("post with sidebar", "sidebar or footer", ...), as in that context the sidebar is a pretty common term.

Curious what others say though.

afercia commented:
I'd agree with @swissspidy that the only occurrences to change should be the ones in the editor UI.

im3dabasia committed:
revert: Panel changes in non editor packages

im3dabasia commented:
Hi @afercia and @swissspidy,
I believe the changes have been made, and the PR is now ready. When you have a moment, I’d appreciate it if you could check the latest changes. 

Hopefully, it can be merged into trunk soon. Thanks!

swissspidy commented on packages/editor/src/components/post-template/create-new-template-modal.js:
```diff
@@ -126,7 +126,7 @@ export default function CreateNewTemplateModal( { onClose } ) {
 						disabled={ isBusy }
 						help={ __(
-							'Describe the template, e.g. "Post with sidebar". A custom template can be manually applied to any post or page.'
+							'Describe the template, e.g. "Post with panel". A custom template can be manually applied to any post or page.'
```
I think this one we also want to ignore, not change.

im3dabasia replied:
Fixed. I have made the changes.

im3dabasia committed:
fix: Revert changes for post-template sidebar word
